python/cuml/cuml/manifold/spectral_embedding.pyx

```diff
@@ -40,6 +47,52 @@ from cuml.internals.mixins import CMajorInputTagMixin, SparseInputTagMixin
 from cuml.internals.utils import check_random_seed
 
 
+def _convert_to_coo_matrix(A):
+    """
+    Convert various matrix formats to COO format.
+
+    Supports:
+    - scipy sparse matrices (CSR, CSC, COO)
+    - cupy sparse matrices (CSR, CSC, COO)
+    - dense numpy arrays
+    - dense cupy arrays
+
+    Parameters
+    ----------
+    A : array-like or sparse matrix
+        Input matrix in various formats
+
+    Returns
+    -------
+    A_coo : scipy.sparse.coo_matrix or cupyx.scipy.sparse.coo_matrix
+        Matrix in COO format
+    """
+    # Check if it's already in COO format
+    if isinstance(A, (scipy_coo, cupy_coo)):
+        return A
+
+    # Handle scipy sparse matrices
+    if scipy_issparse(A):
+        return A.tocoo()
+
+    # Handle cupy sparse matrices
+    if cupy_issparse(A):
+        return A.tocoo()
```

Comment:
`sort_indices` should guarantee order for CSR/CSC. We should probably update the `extract_knn_graph` function too, maybe in an other PR.

Reply:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.sort_indices.html